pkg/pmem-csi-operator/controller/deployment/controller_driver.go

```diff
@@ -158,6 +158,10 @@ func (d *PmemCSIDriver) reconcileDeploymentChanges(r *ReconcileDeployment, exist
 			d.Spec.PMEMPercentage = existing.Spec.PMEMPercentage
 			updateDeployment = true
 			err = fmt.Errorf("changing %q of a running deployment %q is not allowed", c, d.Name)
+		case api.Labels:
+			d.Spec.Labels = existing.Spec.Labels
+			updateDeployment = true
+			err = fmt.Errorf("changing %q of a running deployment %q is not allowed", c, d.Name)
```

Comment:
@pohly We could use `Deployment.ObjectMeta.Labels` instead of adding it to DeploymentSpec. It's more convenient to the user, treating like other Kubernetes objects. And, Is there any specific reason behind choosing `labels` not editable? It feels to me like a useful feature to make it editable.

Reply:
> We could use Deployment.ObjectMeta.Labels instead of adding it to DeploymentSpec. It's more convenient to the user, treating like other Kubernetes objects.

I'm not aware of any other object where attributes of created objects are copied from somewhere other than the "Spec" field. Do you have an example?

> And, Is there any specific reason behind choosing labels not editable? It feels to me like a useful feature to make it editable.

Just a feeling that this would be potentially dangerous and/or confusing because there will be time periods where some objects have the old labels (like running pods) and some the new labels (the daemonset).

Can you give an example under which circumstances an admin might want to mutate labels?

The label selector in a service is also immutable. This doesn't affect us because our service object doesn't use the custom labels, I'm just mentioning it as an example where others have decided that supporting label changes isn't a good idea.

If you want to discuss this further, can we make this an open issue and in the meantime merge this PR?

Reply:
> I'm not aware of any other object where attributes of created objects are copied from somewhere other than the "Spec" field. Do you have an example?

Sorry, If I couldn't get exactly. But my point was we need not introduce a new spec field for labels, as every Kubernetes objects already have ObjectMeta. Similar to how we use `Deployment.Name` we could use `Deployment.Labels`.

Reply:
> Just a feeling that this would be potentially dangerous and/or confusing because there will be time periods where some objects have the old labels (like running pods) and some the new labels (the daemonset).

Ok, I do not have a strong feeling here. I just felt like to match the behaviour of other Kubernetes objects.

Reply:
> Sorry, If I couldn't get exactly. But my point was we need not introduce a new spec field for labels, as every Kubernetes objects already have ObjectMeta.

Yes, but it's not normally used for derived objects. All of the app controllers allow separate labels for the main object and for the objects that get created for it.

> Similar to how we use Deployment.Name

Which is also a bit unusual...

Reply:
> Sorry, If I couldn't get exactly. But my point was we need not introduce a new spec field for labels, as every Kubernetes objects already have ObjectMeta.
> 
> Yes, but it's not normally used for derived objects. All of the app controllers allow separate labels for the main object and for the objects that get created for it.

@pohly Just for reference, can you please give some example API which uses separate labels other than metadata.labels. Kubernetes mentions labels & annotations as [common features](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#common-features) for CRs.

Reply:
That page just says that CRs automatically have labels. It does not say that objects created by an operator inherit the labels.

Example APIs:
- statefulset: labels of pods come from spec.template.metadata.labels, not from metadata.labels; labels of PVCs come from spec.template.metadata.labels
- deployment: labels of the replicaset and pods come from spec.template.metadata.labels, not from metadata.labels

Do you have an example of an API where the labels are copied from the CR's metadata.labels? Which one do we want to follow, core Kubernetes behavior or some add-on?

Reply:
> Example APIs: ...

My point is that these APIs show that objects that describe other objects do so in their "spec" field, not with their own meta data. My gut feeling is that this is a useful, more flexible design.

I know that we don't follow that with the namespace, but that's for technical reasons. The Kubernetes deployment also uses the same namespace even if something else is specified in the template; I haven't tested statefulset, but it probably behaves the same.

I know that we reuse metadata.name as base for created objects. I see that as a shortcut; we might want to add a optional spec.Name field for the sake of consistency.

Reply:
> That page just says that CRs automatically have labels. It does not say that objects created by an operator inherit the labels.
> Example APIs:
> 
> statefulset: labels of pods come from spec.template.metadata.labels, not from metadata.labels; labels of PVCs come from spec.template.metadata.labels
> deployment: labels of the replicaset and pods come from spec.template.metadata.labels, not from metadata.labels
> 

Thanks, @pohly, for pointing examples. Yes, my thinking about labels, in this case, is wrong. I agree let's keep them in spec.